src/Statistics.jl

```diff
@@ -0,0 +1,137 @@
+"""
+function mean!(M::Manifold, y, x::AbstractVector, w::AbstractWeights = weights(ones(length(x)) / length(x));
+            stop_iter=100,
+            kwargs...
+        ) where {T}
+    yold = similar_result(M, mean, y)
+    copyto!(yold,y)
+    v = zero_tangent_vector.(Ref(M), fill(y,length(x)))
+    for i=1:stop_iter
+        copyto!(yold,y)
+        log!.(Ref(M), v, Ref(yold), x)
+        exp!(M, y, yold, sum( values(w).*v ) / 2 )
+        isapprox(M,y,yold; kwargs...) && break
+    end
+    return y
+end
+
+@doc doc"""
+    median(M,x weights=1/n*ones(n); stop_iter=10000, use_rand = false )
+
+computes the Riemannian median of the vector `x`  of `n` points on the
+[`Manifold`](@ref) `M`. This function is nonsmooth (i.e nondifferentiable) and
+uses a cyclic proximal point scheme. Optionally one can provide
+weights $w_i$ for the weighted Riemannian median. The general formula to compute
+the minimizer reads
+````math
+\argmin_{y\in\mathcal M}\sum_{i=1}^n w_i\mathrm{d}_{\mathcal M}(y,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemannian [`distance`](@ref).
+
+The cyclic proximal point can be run with a random cyclic order by setting
+`use_rand` to `true`
+
+Optionally you can provide `x0`, the starting point (by default set to the first
+data point). `stop_iter` denotes the maximal number of iterations to perform and
+the `kwargs...` are passed to [`isapprox`](@ref) to stop, when the inimal change
+between two iterates is small. For more stopping criteria
+check the [`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in Algorithm 4.3 and 4.4 in 
+> Bačák, M: Computing Medians and Means in Hadamard Spaces.
+> SIAM Journal on Optimization (2014), 24(3), pp. 1542–1566,
+> doi: [10.1137/140953393](https://doi.org/10.1137/140953393),
+> arxiv: [1210.2145](https://arxiv.org/abs/1210.2145)
+"""
+function median(M::Manifold, x::AbstractVector, w::AbstractWeights = weights(ones(length(x)) / length(x)); x0=x[1], kwargs...)
+    y = similar_result(M, median, x[1])
+    copyto!(y,x0)
+    return median!(M, y, x; kwargs...)
+end
+@doc doc"""
+    median!(M,y,x,w)
+
+computes the [`median`](@ref) in-place in `y` where the initial value of `y` is
+the starting point of the algorithm.
+"""
+function median!(M::Manifold, y, x::AbstractVector, w::AbstractWeights = weights(ones(length(x)) / length(x));
+            stop_iter=10000,
+            use_random = false,
+            kwargs...
+        ) where {T}
+    n = length(x)
+    yold = similar_result(M,median,y)
+    copyto!(yold,y)
+    order = collect(1:n)
+    iter = 0
+    v = zero_tangent_vector(M,y)
+    for i=1:stop_iter
+        λ = 1/(iter+1)
+        copyto!(yold,y)
+        use_random && shuffle!(order)
+        zero_tangent_vector!(M,v,y)
+        for i=1:n
+            t = min( λ * w[order[i]] / distance(M,y,x[order[i]]) , 1 )
+            log!(M, v, y, x[order[i]])
+            exp!( M, y, y, t* v)
```

Comment:
I think we can't assume that overwriting the point at which the vector is tangent is safe. I'd suggest adding another temporary besides `yold`.

Reply:
Oh that is actually fine. Line 115 computes a new `v`, which is only once used in line 116 to do the step, `y`is updated and directly after that (next loop) `v` is updated in line 115 again.

Reply:
What I mean is that `exp!` should be able to assume that writing to the output doesn't modify any arguments. For example this should work:
```julia
function exp!(S::Sphere, y, x, v)
    θ = norm(S, x, v)
    fill!(y, 0)
    y .= cos(θ) .* x .+ usinc(θ) .* v
    return y
end
```
And it wouldn't called like this (`exp!( M, y, y, t* v)`).

Reply:
Ah and if you pass to both `x` and `y` the same variable you get something strange, yes; hm, so I have to use the non-mutating one here?

Reply:
Found the culprit, I mixed up loop indices at some point, too; but then with the non-mutating `exp` this again works.

Reply:
Yes, a non-mutating `exp` is correct here (although a mutating one with another temporary would be faster).

Reply:
for now I would like to stay with this one and if we need the mutating one (which would not work for one of your manifolds) we could still implement the mutating one with an optional parameter.